### DISCUSSION TITLE
OpenDDS and TSN

### DISCUSSION THREAD

**Discussion by 1GiLu1** [OpenDDS/OpenDDS#4780] [Q&A] [answered]
Hello,

I'm trying to use OpenDDS in combination with Time Sensitive Networking (TSN) under linux. 
Lets say i have an Application, which has two DataWriters A and B. DataWriter A writes low-priority data and its data should be mapped to traffic class 0 of my TSN-network. DataWriter B writes high-priority data and its data should be mapped to traffic class 2 of my TSN-network. A scheduler (Time Aware Shaper) then decides at which times a traffic class is allowed to send data to the TSN-network.

In a raw application in Linux, we can use the SO_PRIORITY flag when creating a socket. With this flag we can map the socket to a TSN traffic class later (in linux with tc qdisc). Is there a way in OpenDDS to configure this SO_PRIORITY  flag, when it creates the sockets?
First i thought, that the Transport Priority QoS would look helpfull, but i guess it don't set the socket priority.

Thanks a lot and have a nice weekend !

**Comment by sonndinh** [comment]
AFAIK, OpenDDS doesn't support mapping from the TRANSPORT_PRIORITY QoS to the underlying transport's priority. I'm not totally sure, though, whether you can specifically configure the socket on Linux using OpenDDS; I'm leaning more toward it isn't supported yet.

**Reply by jrw972** [reply, marked as answer]
FWIW, there is a spec for DDS with TSN being developed https://www.omg.org/spec/DDS-TSN/1.0/Beta1/About-DDS-TSN.
In general, your idea about using the transport priority is good but I think there are some issues.
I'm going to assume you are using the RTPS transport.

First, since this is a socket-level option and each transport manages a single socket, you are looking at a multi-transport situation; one for each traffic class.
The RTPS transport configuration could be extended to take a priority and set the SO_PRIORITY flag accordingly.
In the application, you would then bind each writer/reader to a transport with the correct priority.
The DevGuide has examples for doing this.

One unknown is using multiple RTPS transports.
This may be supported right now or it might need some work.

So, if you want to get this working I would
1.  Test multiple RTPS transports.
2. Add a new configuration option.
3. Implement the code when the socket is opened to set the priority.

**Reply by 1GiLu1** [reply]
Thanks a lot for the fast replies,

I will have a deeper look at how much work that will be.

Right now it's a workaround for me, to configure a transport configuration and set source ip and port there explicitly. Then under linux i can use iptables to  give a priority to every udp-packet with this specified source port. This has the same effect as setting the SO_PRIORTIY on the socket.

This seems to work in a first test. but leads me to another question:

You said, that every transport has its own socket. So if 2 DataWriters use the same transport, they also use the same socket (=same port). So if i want the DataWriters to use its own udp-ports, i have to create two transport configs (in config.ini) and assign them manually to the DataWriters in Code, right?
(As described here: https://opendds.readthedocs.io/en/dds-3.27/devguide/run_time_configuration.html#using-multiple-configurations)

**Reply by jrw972** [reply]
Correct.
